add context.Int to cli context

griechel opened the pull request:
@nmiyake

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/palantir/pkg/40)
<!-- Reviewable:end -->

nmiyake commented:
Minor feedback

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[cli/context.go, line 45 at r1](https://reviewable.io:443/reviews/palantir/pkg/40#-Ki6tvqwtJBEzaigH_JJ:-Ki6tvqxMlGM9MVFH6iD:b-2xahqk) ([raw file](https://github.com/palantir/pkg/blob/951e95a819a303709a3d5d8420123b3f437deae1/cli/context.go#L45)):*
> ```Go
> 
> func (ctx *Context) Int(name string) int {
> 	value, err := strconv.Atoi(ctx.get(name).(string))
> ```

Use `ctx.String(name)` in the call instead of duplicating the logic

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/pkg/40#-:-Ki6uBwqpZgSlWZQ5zv4:b-oh3hfv)*
<!-- Sent from Reviewable.io -->

bmoylan commented on cli/context.go:
```diff
@@ -40,6 +41,14 @@ func (ctx *Context) Slice(name string) []string {
+func (ctx *Context) Int(name string) int {
+	value, err := strconv.Atoi(ctx.get(name).(string))
+	if err != nil {
+		panic(fmt.Errorf("interface conversion: interface is string, not int"))
```
I don't really like panicking on user input, would it be too weird to return an error for Int if the others don't?

griechel replied:
All the other ones panic too if casting goes awry

nmiyake commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[cli/context.go, line 47 at r1](https://reviewable.io:443/reviews/palantir/pkg/40#-Ki6qDANc4ng_9brbzTz-r1-47:-Ki6uzvVkLjth2Z-XiU1:b-x80thp) ([raw file](https://github.com/palantir/pkg/blob/951e95a819a303709a3d5d8420123b3f437deae1/cli/context.go#L47)):*
<details><summary><i>Previously, bmoylan (Brad Moylan) wrote…</i></summary><blockquote>

I don't really like panicking on user input, would it be too weird to return an error for Int if the others don't?
</blockquote></details>

Yeah, these functions are meant to assert as part of the casting -- all of the other ones panic as well.

However, you could probably argue that, if we go this route, we should introduce `Int` as a top-level flag type rather than as a side-effect of parsing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/pkg/40)*
<!-- Sent from Reviewable.io -->

griechel commented:
@nmiyake addressed comments. Good point about the flag. Is that something that I can add in a later PR? This small change is blocking another project's work at the moment

bmoylan commented:
If you don't want to implement `flag.IntFlag` can't you just do the strconv in your application code until then? This isn't _really_ blocking...

---

Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/pkg/40#-:-Ki6xTUA3GbLu4okw103:b-vy4tu3)*
<!-- Sent from Reviewable.io -->

nmiyake commented:
Brad beat me to it -- I have the exact same feeling. `pkg` is very general-purpose and has a higher quality bar -- we want to be sure that things that are added here are conceptually consistent.

Agree that, if we want to semantically declare that a particular flag must store an integer value, then we should do the full change and add an `IntFlag` and propagate that all the way through.

If we don't do that, then your application is simply declaring by convention that it expects the content of particular flags to be `int` values. If that's the case, I would like that logic to live at the application-level rather than in this package. It's only a few lines of code and, as far as I can tell, doesn't depend on any internal/private state, so it should be possible to have on the client-side.

---

Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/pkg/40#-:-Ki6xjD0lchIOgOHQoDN:bwbymh9)*
<!-- Sent from Reviewable.io -->

griechel commented:
You're right, there is definitely a workaround. I'll do that and re-open a PR that does it right and includes the IntFlag

jonsyu1 commented:
I agree with Brad

On Apr 19, 2017, at 1:45 PM, Brad Moylan <notifications@github.com<mailto:notifications@github.com>> wrote:


If you don't want to implement flag.IntFlag can't you just do the strconv in your application code until then? This isn't really blocking...

________________________________

Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions.

________________________________

Comments from Reviewable[reviewable.io]<https://urldefense.proofpoint.com/v2/url?u=https-3A__reviewable.io-3A443_reviews_palantir_pkg_40-23-2D-3A-2DKi6xTUA3GbLu4okw103-3Ab-2Dvy4tu3&d=DwMFaQ&c=izlc9mHr637UR4lpLEZLFFS3Vn2UXBrZ4tFb6oOnmz8&r=b6UG3qRebLXmvfXiVSoNxg&m=kP7mH5OgT3WINI2TdoAv2VSVma-x0IErH8LZo0gPAVg&s=frs4-H00d0qX_p9WrXvAIip5z8uOC-XhD6ctmc8DVTQ&e=>

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub[github.com]<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_palantir_pkg_pull_40-23issuecomment-2D295433956&d=DwMFaQ&c=izlc9mHr637UR4lpLEZLFFS3Vn2UXBrZ4tFb6oOnmz8&r=b6UG3qRebLXmvfXiVSoNxg&m=kP7mH5OgT3WINI2TdoAv2VSVma-x0IErH8LZo0gPAVg&s=VrRBSMmc8KHZ3VhUa_D9Qhz5cPzXcmEkUh-km40jBKc&e=>, or mute the thread[github.com]<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAuyl4fU-2DtxoanZ0jJPWHx6f98-2DV2gE3ks5rxnJwgaJpZM4NCLzT&d=DwMFaQ&c=izlc9mHr637UR4lpLEZLFFS3Vn2UXBrZ4tFb6oOnmz8&r=b6UG3qRebLXmvfXiVSoNxg&m=kP7mH5OgT3WINI2TdoAv2VSVma-x0IErH8LZo0gPAVg&s=5HnAkZthZy0YAkMc9S9pUCRnNsxVyNufv7e7gnb5dzI&e=>.
